### PR TITLE
towards adding Pereira2018ROI as a benchmark

### PR DESCRIPTION
**Pull request by aalok-sathe**
the data file is local for now; not available on s3. We'll share the csv in slack.
this contains means of ROIs for Lang for all subjects, averaged across sessions.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -136,7 +136,7 @@ dmypy.json
 
 ### project specific additions:
 
-brainscore_language/data
+# brainscore_language/data
```
Comment (mschrimpf) [inline .gitignore:139]: dead specification, either delete or leave, no commented code

File: brainscore_language/benchmarks/pereira2018_v2022/__init__.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+  title={Toward a universal decoder of linguistic meaning from brain activation},
+  author={Pereira, Francisco and Lou, Bin and Pritchett, Brianna and Ritter, Samuel and Gershman, Samuel J 
+          and Kanwisher, Nancy and Botvinick, Matthew and Fedorenko, Evelina},
+  journal={Nature communications},
+  volume={9},
+  number={1},
+  pages={1--13},
+  year={2018},
+  publisher={Nature Publishing Group}
+}"""
```
Comment (mschrimpf) [inline brainscore_language/benchmarks/pereira2018_v2022/__init__.py:23]: @kvfairchild we're beginning to duplicate this `BIBTEX` quite a lot, e.g. in [`data/pereira2018`](https://github.com/brain-score/language/blob/b03ea2c9ceef646c53a7caedcab9ac13df07354a/brainscore_language/data/pereira2018/__init__.py#L8-L18) and [`benchmarks/pereira2018`](https://github.com/brain-score/language/blob/b03ea2c9ceef646c53a7caedcab9ac13df07354a/brainscore_language/benchmarks/pereira2018/__init__.py#L13-L23). One simple solution might be to replace the `benchmarks` definitions with `from brainscore_language.data.pereira2018 import BIBTEX`, is that acceptable from a plugin perspective? I guess we might run into import errors if dependencies are not installed and potentially encourage direct imports when we want plugins to go through the plugin manager.

File: brainscore_language/benchmarks/pereira2018_v2022/__init__.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
+import logging
```
Comment (mschrimpf) [inline brainscore_language/benchmarks/pereira2018_v2022/__init__.py:1]: Shouldn't this just replace the current `pereira2018`? I would say we don't want a new file, but rather want to make this the new version.
Tagging @kvfairchild so she can chime in from the plugins perspective.

File: brainscore_language/benchmarks/pereira2018_v2022/__init__.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+        rather than testing every layer separately
+
+    Each of these benchmarks evaluates one of the two experiments, the overall Pereira2018-linear score is the mean of
+    the two ceiling-normalized scores.
+    """
+
+    def __init__(
+        self,
+        experiment: str,
+        # ceiling_s3_kwargs: dict = None,
```
Comment (mschrimpf) [inline brainscore_language/benchmarks/pereira2018_v2022/__init__.py:54]: dead code; put on S3

File: brainscore_language/benchmarks/pereira2018_v2022/__init__.py [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+            regression_kwargs={
+                "stimulus_coord": "sentence",
+            },
+            correlation_kwargs={
+                "correlation_coord": "sentence",
+                "neuroid_coord": "neuroid",
+            },
+        )
+        identifier = f"Pereira2018.{experiment}-linreg_pearsonr"
+        # ceiling = self._load_ceiling(identifier=identifier, **ceiling_s3_kwargs)
```
Comment (mschrimpf) [inline brainscore_language/benchmarks/pereira2018_v2022/__init__.py:71]: dead code

File: brainscore_language/benchmarks/pereira2018_v2022/__init__.py [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+        super(_Pereira2018LinregPearsonr, self).__init__(
+            identifier=identifier,
+            version=1,
+            parent="Pereira2018-linear",
+            ceiling_func=None,  # ceiling,
```
Comment (mschrimpf) [inline brainscore_language/benchmarks/pereira2018_v2022/__init__.py:76]: dead code

File: brainscore_language/benchmarks/pereira2018_v2022/__init__.py [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+            bibtex=BIBTEX,
+        )
+
+    def _load_data(self, experiment: str) -> NeuroidAssembly:
+        data = load_dataset("Pereira2018_v2022.language")
+        # data = data.sel(experiment=experiment)  # filter experiment
```
Comment (mschrimpf) [inline brainscore_language/benchmarks/pereira2018_v2022/__init__.py:82]: dead code

File: brainscore_language/benchmarks/pereira2018_v2022/__init__.py [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+    #         sha1=sha1,
+    #     )
+    #     if raw_kwargs:  # recursively load raw attributes
+    #         raw = self._load_ceiling(
+    #             identifier=identifier,
+    #             assembly_prefix=assembly_prefix + "raw_",
+    #             **raw_kwargs,
+    #         )
+    #         ceiling.attrs["raw"] = raw
+    #     return ceiling
```
Comment (mschrimpf) [inline brainscore_language/benchmarks/pereira2018_v2022/__init__.py:112]: lots of dead code

File: brainscore_language/benchmarks/pereira2018_v2022/__init__.py [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+            recording_type=ArtificialSubject.RecordingType.fMRI,
+        )
+        stimuli = self.data["sentence"]
+        predictions = candidate.digest_text(stimuli.values)["neural"]
+        predictions["presentation"] = "presentation", stimuli["sentence"].values
+        raw_score = self.metric(
+            predictions,
+            self.data,
+        )
+        score = raw_score  # ceiling_normalize(raw_score, self.ceiling)
```
Comment (mschrimpf) [inline brainscore_language/benchmarks/pereira2018_v2022/__init__.py:126]: dead code

File: brainscore_language/benchmarks/pereira2018_v2022/ceiling_packaging.py [hunk 10]
```diff
@@ -0,0 +1,305 @@
+import logging
```
Comment (mschrimpf) [inline brainscore_language/benchmarks/pereira2018_v2022/ceiling_packaging.py:1]: this entire file seems to be a copy of `pereira2018/ceiling_packaging.py`, I'm not sure what it's doing

File: brainscore_language/benchmarks/pereira2018_v2022/test.py [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+
+        # TODO: we are lazy-dropping benchark.data timeid dim, but drop in packaging or accommodate donwstream later?
+        benchmark.data = benchmark.data.squeeze(dim="time", drop=True)
+
+        exact_data = copy.deepcopy(benchmark.data)  # .reset_index("presentation")
+        # exact_data = exact_data.drop_vars("sentence")
+        del (exact_data["presentation"],)
+        # exact_data = exact_data.set_index(
+        #     presentation=list(exact_data["presentation"].coords)
+        # )
```
Comment (mschrimpf) [inline brainscore_language/benchmarks/pereira2018_v2022/test.py:78]: dead code

File: brainscore_language/benchmarks/pereira2018_v2022/test.py [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+    #     assert hasattr(ceiling, "raw")
+    #     assert set(ceiling.raw.dims) == {"neuroid"}
+    #     assert ceiling.raw.median() == ceiling
+    #     assert hasattr(ceiling.raw, "raw")
+    #     assert set(ceiling.raw.raw.dims) == {
+    #         "sub_subject",
+    #         "num_subjects",
+    #         "split",
+    #         "neuroid",
+    #     }
```
Comment (mschrimpf) [inline brainscore_language/benchmarks/pereira2018_v2022/test.py:109]: dead code

File: brainscore_language/benchmarks/pereira2018_v2022/test.py [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+
+
+if __name__ == "__main__":
+    test = TestBenchmark()
+    test.test_exact(243)
```
Comment (mschrimpf) [inline brainscore_language/benchmarks/pereira2018_v2022/test.py:114]: no need for this, instead run with pytest

File: brainscore_language/data/pereira2018_v2022/__init__.py [hunk 14]
```diff
@@ -0,0 +1,23 @@
+import logging
+
+from brainscore_language import data_registry
+
+# from brainscore_language.utils.s3 import load_from_s3
```
Comment (mschrimpf) [inline brainscore_language/data/pereira2018_v2022/__init__.py:5]: dead code

File: brainscore_language/data/pereira2018_v2022/__init__.py [hunk 15]
```diff
@@ -0,0 +1,23 @@
+import logging
```
Comment (mschrimpf) [inline brainscore_language/data/pereira2018_v2022/__init__.py:1]: Let's talk about whether this goes into its own file or is part of the existing pereira2018 data plugin (I put a note on the agenda)

File: brainscore_language/data/pereira2018_v2022/data_packaging.py [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+        zip(set(data["Sentence"]), np.arange(len(set(data["Sentence"])))),
+    )
+
+    presentation_arr = [None] * len(presentation_id_map)
+    experiment_arr = [None] * len(presentation_id_map)
+    neuroid_arr = [None] * len(neuroid_id_map)
+    subject_arr = [None] * len(neuroid_id_map)
+
+    # inv_neuroid_map = {v: k for k, v in neuroid_id_map}
+    # inv_presentation_map = {v: k for k, v in presentation_id_map}
```
Comment (mschrimpf) [inline brainscore_language/data/pereira2018_v2022/data_packaging.py:53]: feeling less strongly about it here, but still: dead code :)

File: brainscore_language/data/pereira2018_v2022/data_packaging.py [hunk 17]
```diff
@@ -0,0 +1,120 @@
...
+    )
+    s3_key = netcdf_file_name
+
+    # write to disk and upload
+    netcdf_kf_sha1 = write_netcdf(assembly, target_netcdf_path)
+    _logger.debug(f"Wrote file to {target_netcdf_path}.")
+    # upload_to_s3(target_netcdf_path, bucket_name, s3_key)
+    # _logger.debug(
+    #     f"Uploaded assembly {assembly_identifier} to S3: {s3_key} (SHA1 hash {netcdf_kf_sha1})"
+    # )
```
Comment (mschrimpf) [inline brainscore_language/data/pereira2018_v2022/data_packaging.py:115]: can just reuse https://github.com/brain-score/language/blob/b03ea2c9ceef646c53a7caedcab9ac13df07354a/brainscore_language/utils/s3.py#L12

File: brainscore_language/data/pereira2018_v2022/test.py [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+        _logger.info(
+            f'stimulus "The concert pianist went blind in adulthood." is present!'
+        )
+
+        assert len(set(assembly["subject"].values)) == 10
+        _logger.info(f'no. of subjects == {len(set(assembly["subject"].values))}')
+
+        assert (
+            len(set(assembly["neuroid"].values)) == 120
+        )  # potential TODO: rename to neuroid_id
```
Comment (mschrimpf) [inline brainscore_language/data/pereira2018_v2022/test.py:43]: todo?

**Review by mschrimpf — CHANGES_REQUESTED** [review]
High-level comment: I think the benchmark should go inside `pereira2018` rather than a new plugin `pereira2018_v2022` -- @kvfairchild do you agree?

Note that unit tests are failing; if you'd like to put this data onto the S3 bucket that we used for the other data, slack me and I'll get you write access.

**Comment by mschrimpf** [comment]
@aalok-sathe any updates on this PR?

**Comment by aalok-sathe** [comment]
should we be concerned by this (warning?) message?
`DEBUG:brainscore_core.metrics:isel on raw values failed: ValueError("Dimensions {'subject'} do not exist. Expected one or more of ('split', 'neuroid')")
`

we encounter it at https://github.com/brain-score/language/blob/b2f24265b4af0470a27fad899ef5391e244dd5ed/brainscore_language/utils/transformations.py#L232 during ceiling computation and aren't sure if this is an issue.
our assembly contains `subject` as a coord but not as a dimension (subject is never a dimension, right? it is along the `neuroid` dimension?)

**Comment by aalok-sathe** [comment]
We're also looking for some help towards this failure of the test; now that we uploaded to s3 we're not sure why it's failing to find the assembly file in the cache: https://app.travis-ci.com/github/brain-score/language/builds/255739587#L660

**Comment by mschrimpf** [comment]
> should we be concerned by this (warning?) message? `DEBUG:brainscore_core.metrics:isel on raw values failed: ValueError("Dimensions {'subject'} do not exist. Expected one or more of ('split', 'neuroid')") `

No -- this is just saying that it cannot select on the score's `raw` values, so it is not selecting anything there. Maybe we should turn this into a debug message..

> We're also looking for some help towards this failure of the test; now that we uploaded to s3 we're not sure why it's failing to find the assembly file in the cache: https://app.travis-ci.com/github/brain-score/language/builds/255739587#L660

At least in the build you linked, it's trying to load the file from local disk rather than S3: 
```
brainscore_language/data/pereira2018_v2022/__init__.py:23: in <lambda>
    data_registry["Pereira2018_v2022.language"] = lambda: load_from_disk("Pereira2018ROI")
```

File: tests/test_test_plugins.py [hunk 19]
```diff
@@ -15,12 +15,12 @@
...
-		Path(DUMMY_REQUIREMENTS).touch()
-		Path(DUMMY_TESTFILE).touch()
-		with open(DUMMY_TESTFILE, 'w') as f:
-			f.write(textwrap.dedent('''\
+    def setup_method(self):
+        DUMMY_PLUGIN_PATH.mkdir(parents=True, exist_ok=True)
+        Path(DUMMY_REQUIREMENTS).touch()
+        Path(DUMMY_TESTFILE).touch()
+        with open(DUMMY_TESTFILE, 'w') as f:
+            f.write(textwrap.dedent('''\
```
Comment (mschrimpf) [inline tests/test_test_plugins.py:23]: fyi this is leading to errors : `TabError: inconsistent use of tabs and spaces in indentation`

**Comment by aalok-sathe** [comment]
> > We're also looking for some help towards this failure of the test; now that we uploaded to s3 we're not sure why it's failing to find the assembly file in the cache: https://app.travis-ci.com/github/brain-score/language/builds/255739587#L660
> 
> At least in the build you linked, it's trying to load the file from local disk rather than S3:
> 
> ```
> brainscore_language/data/pereira2018_v2022/__init__.py:23: in <lambda>
>     data_registry["Pereira2018_v2022.language"] = lambda: load_from_disk("Pereira2018ROI")
> ```

I think we changed that in the latest commit we pushed yesterday:
https://github.com/brain-score/language/commit/a74399462a68102dd8545c3aa57a6d29e3a946c6

File: brainscore_language/data/pereira2018_v2022/test.py (hunk 18 above)
Comment (gretatuckute) [reply brainscore_language/data/pereira2018_v2022/test.py:43]: yes to do

File: brainscore_language/data/pereira2018_v2022/__init__.py [hunk 20]
```diff
@@ -0,0 +1,28 @@
...
+  publisher={Nature Publishing Group}
+}"""
+
+
+# data_registry["Pereira2018_v2022.language"] = lambda: load_from_disk("Pereira2018ROI")
+data_registry["Pereira2018_v2022.language"] = lambda: load_from_s3(
+    identifier="Pereira2018ROI",
+    version_id="8kEzAXaWnxjMIC95MrhSJ3gtP2a1W.Y7",
+    sha1="235ecf1f2abc38cb892c0a7b5c9c465434f153c5",
+)
```
Comment (mschrimpf) [inline brainscore_language/data/pereira2018_v2022/__init__.py:28]: @aalok-sathe I set the S3 object to public, but I think the issue is that you're attempting to load `assy_Pereira2018ROI` while the S3 file is called `assembly_Pereira2018ROI` -- if you pass `assembly_prefix='assembly'` to https://github.com/brain-score/language/blob/5ea01ce9f619ca23bcce9c54da6c47669def1138/brainscore_language/utils/s3.py#L33, I believe it should work

**Comment by gretatuckute** [comment]
> > should we be concerned by this (warning?) message? `DEBUG:brainscore_core.metrics:isel on raw values failed: ValueError("Dimensions {'subject'} do not exist. Expected one or more of ('split', 'neuroid')") `
> 
> No -- this is just saying that it cannot select on the score's `raw` values, so it is not selecting anything there. Maybe we should turn this into a debug message..
> 
> > We're also looking for some help towards this failure of the test; now that we uploaded to s3 we're not sure why it's failing to find the assembly file in the cache: https://app.travis-ci.com/github/brain-score/language/builds/255739587#L660
> 
> At least in the build you linked, it's trying to load the file from local disk rather than S3:
> 
> ```
> brainscore_language/data/pereira2018_v2022/__init__.py:23: in <lambda>
>     data_registry["Pereira2018_v2022.language"] = lambda: load_from_disk("Pereira2018ROI")
> ```

It already shows up as a DEBUG message. But I don't think we understand why it would occur there in the first place? In tracks to a brainio script where there is even no "isel" operation?

**Review by gretatuckute — APPROVED** [review]

**Comment by mschrimpf** [comment]
Sorry, this PR was closed after combining commit history. @aalok-sathe can you please open a new PR?

**Comment by aalok-sathe** [comment]
No worries! will do!

________________________________

**Pull request closed without merging**